miden-crypto/src/merkle/forest/mod.rs

```diff
@@ -0,0 +1,176 @@
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an opening for the specified key in the SMT with the specified root.
+    ///
+    /// Returns an error if an SMT with this root is not in the forest, or if the forest does
+    /// not have sufficient data to provide an opening for the specified key.
+    pub fn open(&self, root: Word, key: Word) -> Result<SmtProof, MerkleError> {
+        if !self.roots.contains(&root) {
+            return Err(MerkleError::RootNotInStore(root));
+        }
+
+        let index = key[3].as_int();
+        let node_index = NodeIndex::new(SMT_DEPTH, index)?;
+        let path: SparseMerklePath = self.store.get_path(root, node_index)?.path.try_into()?;
+
+        let leaf = match self.leaves.get(&key) {
+            Some(leaf) => leaf.clone(),
+            None => return Err(MerkleError::UntrackedKey(key)),
+        };
+
+        Ok(match SmtProof::new(path, leaf) {
+            Ok(proof) => proof,
+            Err(err) => match err {
+                SmtProofError::InvalidMerklePathLength(depth) => {
+                    return Err(MerkleError::InvalidPathLength(depth));
+                },
+            },
+        })
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Inserts the specified key-value pair into an SMT with the specified root. This would also
+    /// add a new root to the forest.
+    ///
+    /// Returns an error if an SMT with the specified root is not in the forest, these is not
+    /// enough data in the forest to perform the insert, or if the insert would create a leaf
+    /// with too many entries.
+    pub fn insert(&mut self, root: Word, key: Word, value: Word) -> Result<Word, MerkleError> {
+        if !self.roots.contains(&root) {
+            return Err(MerkleError::RootNotInStore(root));
+        }
+
+        let index = key[3].as_int();
+        let node_index = NodeIndex::new(SMT_DEPTH, index)?;
+        let path = self.store.get_path(root, node_index)?.path;
+
+        let leaf_hash = match self.leaves.get_mut(&key) {
+            Some(leaf) => match leaf.insert(key, value) {
+                Ok(_) => leaf.hash(),
+                Err(err) => match err {
+                    SmtLeafError::TooManyLeafEntries { actual } => {
+                        return Err(MerkleError::TooManyEntries((actual)));
+                    },
+                    _ => {
+                        unreachable!("other SmtLeafError variants should not be possible here");
+                    },
+                },
+            },
+            None => {
+                let leaf = SmtLeaf::new_single(key, value);
+                let leaf_hash = leaf.hash().clone();
+                self.leaves.insert(key, leaf);
+                leaf_hash
+            },
+        };
+        let new_root = self.store.add_merkle_path(index, leaf_hash, path)?;
```

Comment:
You could skip validations altogether by adding a new "unauthenticated_add" method to `MerkleStore` that inserts the new leaf without merkle path authentication, and adds all new hashes on the path from the leaf to root.

Similar approach could be used for batch_insert. You blindly add all new leaves, and process them upwards in BFS order.